upgrade/v1beta1/upgrade.go

```diff
@@ -0,0 +1,287 @@
+	upgradeLogPrefix = "upgrade from v1beta1 to v1beta2:"
+)
+
+func UpgradeCRFromV1beta1ToV1beta2(config *restclient.Config, namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+" failed")
+	}()
+
+	if err := upgradeVolumes(namespace, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeEngineImages(namespace, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeBackupTargets(namespace, lhClient); err != nil {
+		return err
+	}
+	if err := upgradeNodes(namespace, lhClient); err != nil {
+		return err
+	}
+
+	logrus.Infof("%v: completed", upgradeLogPrefix)
+	return nil
+}
+
+func CanUpgrade(config *restclient.Config, namespace string) (bool, error) {
+	lhClient, err := lhclientset.NewForConfig(config)
+	if err != nil {
+		return false, errors.Wrap(err, "unable to get clientset for v1beta1")
+	}
+
+	scheme := runtime.NewScheme()
+	if err := longhornV1beta1.SchemeBuilder.AddToScheme(scheme); err != nil {
+		return false, errors.Wrap(err, "unable to create scheme for v1beta1")
+	}
+
+	setting, err := lhClient.LonghornV1beta1().Settings(namespace).Get(context.TODO(), string(types.SettingNameDefaultEngineImage), metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("setting %v not found", string(types.SettingNameDefaultEngineImage))
+			return true, nil
+		}
+
+		return false, errors.Wrap(err, fmt.Sprintf("unable to get setting %v", string(types.SettingNameDefaultEngineImage)))
+	}
+
+	return false, errors.Wrapf(err, "unable to upgrade from %v directly", setting.Value)
+}
+
+func upgradeVolumes(namespace string, lhClient *lhclientset.Clientset) error {
+	volumes, err := lhClient.LonghornV1beta1().Volumes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return errors.Wrapf(err, "unable to list volumes (%v)", apierrors.ReasonForError(err))
+	}
+
+	for _, old := range volumes.Items {
+		if len(old.Status.Conditions) == 0 {
+			continue
+		}
+
+		new := longhorn.Volume{}
+
+		if err := copier.Copy(&new, &old); err != nil {
+			return err
+		}
+
+		recurringJobs, err := copyRecurringJobs(old.Spec.RecurringJobs)
+		if err != nil {
+			return errors.Wrap(err, "failed to copy volume Spec.RecurringJobs")
+		}
+		new.Spec.RecurringJobs = recurringJobs
+
+		conditions, err := copyConditions(old.Status.Conditions)
+		if err != nil {
+			return errors.Wrap(err, "fail to copy volume Status.Conditions")
+		}
+		new.Status.Conditions = conditions
+
+		if err := copier.Copy(&new.Status.KubernetesStatus, &old.Status.KubernetesStatus); err != nil {
+			return errors.Wrap(err, "failed to copy volume Status.KubernetesStatus")
+		}
+
+		if err := copier.Copy(&new.Status.CloneStatus, &old.Status.CloneStatus); err != nil {
+			return errors.Wrap(err, "failed to copy volume Status.KubernetesStatus")
+		}
+
+		if err := copier.Copy(&new.Status.KubernetesStatus, &old.Status.KubernetesStatus); err != nil {
+			return err
+		}
```

Comment:
Why do need to copy these fields? [1]

Reply:
Because the data type is different, copier.Copy() cannot copy them.

Reply:
But the KubernetesStatus, CloneStatus does not contain the Conditions fields inside.
Also, the status.KubernetesStatus copied twice, intended?

Reply:
> Also, the status.KubernetesStatus copied twice, intended?

Yes, already fixed